README.md
我愛卡提諾
==========

[![PyPI Version](https://badge.fury.io/py/iloveck101.png)](https://pypi.python.org/pypi/iloveck101) [![Build Status](https://travis-ci.org/tzangms/iloveck101.png?branch=master)](https://travis-ci.org/tzangms/iloveck101) [![Coverage Status](https://coveralls.io/repos/tzangms/iloveck101/badge.png)](https://coveralls.io/r/tzangms/iloveck101) [![Bitdeli Badge](https://d2weczhvl823v0.cloudfront.net/tzangms/iloveck101/trend.png)](https://bitdeli.com/free "Bitdeli Badge")

沒錯, 你沒看錯, 我愛卡提諾。

我只想看圖, 可是卡提諾的網站載入太慢, 一堆廣告跟不必要的資訊, 還要等圖片 lazy load, 所以寫了這個 iloveck101, 直接先把圖片都下載下來, 看圖快多了。

目前只支援 Mac, Linux 我不確定。

by tzangms aka 海總理 2013.12.02



如何安裝
==========

```bash
$ sudo pip install iloveck101
```


如何使用
===========

```bash
$ iloveck101 [url]

```

for example, 你可以試試看思穎

```bash
$ iloveck101 http://ck101.com/thread-2876990-1-1.html
```

最後你可以在你 Mac 的「圖片」資料夾裡發現 `iloveck101` 的目錄, 你會很驚訝, 怎麼圖片都已經抓好放在裡面了呢?! 
沒錯, 這就是對卡提諾的愛啊!


看起來像這樣
==============

![圖片資料夾](https://raw.github.com/tzangms/iloveck101/master/docs/images1.png) 
![圖片](https://raw.github.com/tzangms/iloveck101/master/docs/images2.png) 



CHANGES.txt
v0.1.0, 2013.12.02 -- Initial release.
v0.1.1, 2013.12.02 -- Update MANIFEST.in
v0.2.0, 2013.12.03 -- Improvements, Retry 3 times when error, save images to ~/Pictures/iloveck101, URL restriction
v0.3.0, 2013.12.03 -- fix Retry, and folder name with slash error.
v0.4.0, 2013.12.03 -- Remove Pillow dependency, it's too large, and required to compile
v0.5.0, 2013.12.04 -- Add user agent by @kenzouyeh, use `gevent` to speed up by @yhchan
v0.5.1, 2013.12.04 -- Add full user agent
v0.5.2, 2013.12.06 -- Add ability to parse ck101's article list, by @davidyen11244


tests.py
# coding: utf-8

import unittest

from iloveck101.utils import parse_url


class CK101Test(unittest.TestCase):
    def setUp(self):
        self.url = 'http://ck101.com/thread-2818521-1-1.html'

    def test_parse_url(self):
        title, image_urls = parse_url(self.url)
        self.assertEqual(title, u'蔡依林+心凌的綜合體？！19歲「無名時代的正妹」陳敬宣現在正翻了......')
        self.assertEqual(len(image_urls), 39)


if __name__ == '__main__':
    unittest.main()


tests/__init__.py
# coding: utf-8

import os
import unittest

from iloveck101.utils import parse_url, get_image_info
from iloveck101.iloveck101 import iloveck101, main
from iloveck101.exceptions import URLParseError
from httmock import urlmatch, HTTMock, all_requests


TEST_DIR = os.path.abspath(os.path.join(__file__, '..'))


@urlmatch(netloc=r'(.*\.)?imgs\.cc$')
def image_mock(url, request):
    image = os.path.join('fixtures', 'jYY7oMF.jpg')
    with open(os.path.join(TEST_DIR, image)) as f:
        content = f.read()

    return content


@urlmatch(netloc=r'(.*\.)?ck101\.com$')
def thread_mock(url, request):
    html = os.path.join('fixtures', '2818521.html')
    with open(os.path.join(TEST_DIR, html)) as f:
        content = f.read()

    return content


@all_requests
def list_mock(url, request):
    html = os.path.join('fixtures', 'beauty.html')
    with open(os.path.join(TEST_DIR, html)) as f:
        content = f.read()

    return content

@all_requests
def error_mock(url, request):
    return {'status_code': 400, 'content': 'Hi'}



class CK101Test(unittest.TestCase):
    def setUp(self):
        self.thread_url = 'http://ck101.com/thread-2818521-1-1.html'
        self.list_url = 'http://ck101.com/beauty/'

    def test_thread(self):

        with HTTMock(thread_mock, image_mock):
            iloveck101(self.thread_url)


    def test_list(self):
        with HTTMock(list_mock, image_mock):
            iloveck101(self.list_url)

        with HTTMock(error_mock):
            iloveck101(self.list_url)

    def test_not_ck101(self):
        with self.assertRaises(SystemExit) as cm:
            iloveck101('http://google.com/')

        self.assertEqual(cm.exception.code, 'This is not ck101 url')

    def test_retry(self):
        with HTTMock(error_mock):
            with self.assertRaises(SystemExit) as cm:
                iloveck101(self.thread_url)

        self.assertEqual(cm.exception.code, 'Oops, can not fetch the page')

    def test_main(self):
        with self.assertRaises(SystemExit) as cm:
            main()

        self.assertEqual(cm.exception.code, 'This is not ck101 url')


class UtilsTest(unittest.TestCase):
    def setUp(self):
        self.url = 'http://ck101.com/thread-2818521-1-1.html'


    def test_parse_url(self):
        with HTTMock(thread_mock):
            title, image_urls = parse_url(self.url)

        self.assertEqual(title, u'蔡依林+心凌的綜合體？！19歲「無名時代的正妹」陳敬宣現在正翻了......')
        self.assertEqual(len(image_urls), 39)

    def test_parse_url_error(self):
        with HTTMock(error_mock):
            with self.assertRaises(URLParseError):
                title, image_urls = parse_url(self.url)

    def test_get_image_info(self):
        jpg = os.path.join('fixtures', 'jYY7oMF.jpg')

        with open(os.path.join(TEST_DIR, jpg)) as f:
            content_type, width, height = get_image_info(f.read())

        self.assertEqual(content_type, 'image/jpeg')
        self.assertEqual(width, 612)
        self.assertEqual(height, 612)


        png = os.path.join('fixtures', 'firefox.png')

        with open(os.path.join(TEST_DIR, png)) as f:
            content_type, width, height = get_image_info(f.read())

        self.assertEqual(content_type, 'image/png')
        self.assertEqual(width, 256)
        self.assertEqual(height, 256)


        gif = os.path.join('fixtures', 'firefox.gif')

        with open(os.path.join(TEST_DIR, gif)) as f:
            content_type, width, height = get_image_info(f.read())

        self.assertEqual(content_type, 'image/gif')
        self.assertEqual(width, 256)
        self.assertEqual(height, 256)


if __name__ == '__main__':
    unittest.main()


iloveck101/utils.py
import struct
from cStringIO import StringIO

import requests
from lxml import etree

from .exceptions import URLParseError


def get_image_info(data):
    """
    read image dimension
    """

    data = str(data)
    size = len(data)
    height = -1
    width = -1
    content_type = ''

    # handle GIFs
    if (size >= 10) and data[:6] in ('GIF87a', 'GIF89a'):
        # Check to see if content_type is correct
        content_type = 'image/gif'
        w, h = struct.unpack("<HH", data[6:10])
        width = int(w)
        height = int(h)

    # See PNG 2. Edition spec (http://www.w3.org/TR/PNG/)
    # Bytes 0-7 are below, 4-byte chunk length, then 'IHDR'
    # and finally the 4-byte width, height
    elif ((size >= 24) and data.startswith('\211PNG\r\n\032\n')
          and (data[12:16] == 'IHDR')):
        content_type = 'image/png'
        w, h = struct.unpack(">LL", data[16:24])
        width = int(w)
        height = int(h)

    # Maybe this is for an older PNG version.
    elif (size >= 16) and data.startswith('\211PNG\r\n\032\n'):
        # Check to see if we have the right content type
        content_type = 'image/png'
        w, h = struct.unpack(">LL", data[8:16])
        width = int(w)
        height = int(h)

    # handle JPEGs
    elif (size >= 2) and data.startswith('\377\330'):
        content_type = 'image/jpeg'
        jpeg = StringIO(data)
        jpeg.read(2)
        b = jpeg.read(1)
        try:
            while (b and ord(b) != 0xDA):
                while (ord(b) != 0xFF): b = jpeg.read(1)
                while (ord(b) == 0xFF): b = jpeg.read(1)
                if (ord(b) >= 0xC0 and ord(b) <= 0xC3):
                    jpeg.read(3)
                    h, w = struct.unpack(">HH", jpeg.read(4))
                    break
                else:
                    jpeg.read(int(struct.unpack(">H", jpeg.read(2))[0])-2)
                b = jpeg.read(1)
            width = int(w)
            height = int(h)
        except struct.error:
            pass
        except ValueError:
            pass

    return content_type, width, height

def parse_url(url):
    """
    parse image_url from given url
    """

    REQUEST_HEADERS = {
        'User-agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_9_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/31.0.1650.57 Safari/537.36'
    }

    # fetch html and find images
    title = None
    for attemp in range(3):
        resp = requests.get(url, headers=REQUEST_HEADERS)
        if resp.status_code != 200:
            print 'Retrying ...'
            continue

        # parse html
        html = etree.HTML(resp.content)

        # title
        try:
            title = html.find('.//title').text.split(' - ')[0].replace('/', '').strip()
            break
        except AttributeError:
            print 'Retrying ...'
            continue

    if title is None:
        raise URLParseError

    image_urls = html.xpath('//img/@file')
    return title, image_urls


iloveck101/iloveck101.py
import os
import sys
import re
import platform

import gevent
from gevent import monkey
import requests
from lxml import etree
from more_itertools import chunked

from .utils import get_image_info, parse_url
from .exceptions import URLParseError


monkey.patch_all()

REQUEST_HEADERS = {'User-agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_9_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/31.0.1650.57 Safari/537.36'}
BASE_URL = 'http://ck101.com/'
CHUNK_SIZE = 3


def iloveck101(url):
    """
    Determine the url is valid. And check if the url contains any thread link or it's a thread.
    """

    if 'ck101.com' in url:
        if 'thread' in url:
            retrieve_thread(url)
        else:
            for thread in retrieve_thread_list(url):
                if thread is not None:
                    retrieve_thread(thread)
    else:
        sys.exit('This is not ck101 url')


def retrieve_thread_list(url):
    """
    The url may contains many thread links. We parse them out.
    """

    resp = requests.get(url, headers=REQUEST_HEADERS)

    # parse html
    html = etree.HTML(resp.content)

    links = html.xpath('//a/@href')
    for link in links:
        yield link


def retrieve_thread(url):
    """
    download images from given ck101 URL
    """

    # check if the url has http prefix
    if not url.startswith('http'):
        url = BASE_URL + url

    # find thread id
    m = re.match('thread-(\d+)-.*', url.rsplit('/', 1)[1])
    if not m:
        return

    print '\nVisit %s' % (url)

    thread_id = m.group(1)


    # create `iloveck101` folder in ~/Pictures
    system = platform.system()
    if system == 'Darwin':
        picfolder = 'Pictures'
    else:
        picfolder = ''

    home = os.path.expanduser("~")
    base_folder = os.path.join(home, picfolder, 'iloveck101')

    if not os.path.exists(base_folder):
        os.mkdir(base_folder)

    # parse title and images
    try:
        title, image_urls = parse_url(url)
    except URLParseError:
        sys.exit('Oops, can not fetch the page')

    # create target folder for saving images
    folder = os.path.join(base_folder, "%s - %s" % (thread_id, title))
    if not os.path.exists(folder):
        os.mkdir(folder)

    def process_image_worker(image_url):
        filename = image_url.rsplit('/', 1)[1]

        # ignore useless image
        if not image_url.startswith('http'):
            return

        # fetch image
        print 'Fetching %s ...' % image_url
        resp = requests.get(image_url, headers=REQUEST_HEADERS)

        # ignore small images
        content_type, width, height = get_image_info(resp.content)
        if width < 400 or height < 400:
            print "image is too small"
            return

        # save image
        with open(os.path.join(folder, filename), 'wb+') as f:
            f.write(resp.content)

    for chunked_image_urls in chunked(image_urls, CHUNK_SIZE):
        jobs = [gevent.spawn(process_image_worker, image_url)
                for image_url in chunked_image_urls]
        gevent.joinall(jobs)


def main():
    try:
        url = sys.argv[1]
    except IndexError:
        sys.exit('Please provide URL from ck101')

    iloveck101(url)
